### PR TITLE
Polish QualifiedType and related code

### LINKED ISSUE #1395: Polish some API before release
There are 2 things I'd like to improve before we release some of our new API:

image

- the varargs warning on `with(...)`
- the fact that QualifiedType doesn't have a `T` so that a `@Json String` isn't recognized as a `String` and we need to do avoidable casts

### PR DESCRIPTION
**Pull request by TheRealMarnes**
fixes #1395

### CODE REVIEW

**Comment by TheRealMarnes** [comment]
I think we should leave this convenience method out:

image

It's a shorthand for a QualifiedType, but while the varargs warning can be fixed on qualtyp, it can't be fixed here because ResultBearing is an interface and cannot promise safe varargs handling. I don't think this is unreasonable instead:

image

@qualidafial @stevenschlansker ?

**Comment by TheRealMarnes** [comment]
All changes pushed in sequential/semantic commits to clearly justify each change and make it easy to follow. I suggest checking #e28afef on its own, and just checking the global diff for the rest. All the changes relate to QualifiedType so backward compatibility is moot. I've avoided changes to imports too ^^ I think this should be a piece of cake to greenlight.

File: core/src/main/java/org/jdbi/v3/core/mapper/QualifiedColumnMapperFactory.java [hunk 1]
```diff
@@ -13,27 +13,37 @@
...
-    Optional<ColumnMapper<?>> build(QualifiedType type, ConfigRegistry config);
+    <T> Optional<ColumnMapper<T>> build(QualifiedType<T> type, ConfigRegistry config);
 
     static QualifiedColumnMapperFactory adapt(ColumnMapperFactory factory) {
         Set<Annotation> qualifiers = getQualifiers(factory.getClass());
-        return (type, config) -> type.getQualifiers().equals(qualifiers)
-            ? factory.build(type.getType(), config)
-            : Optional.empty();
+
+        return new QualifiedColumnMapperFactory() {
```
Comment (TheRealMarnes) [inline core/src/main/java/org/jdbi/v3/core/mapper/QualifiedColumnMapperFactory.java]: this and the other inline class are necessary to get a T reference to work with

File: core/src/main/java/org/jdbi/v3/core/mapper/reflect/BeanMapper.java [hunk 2]
```diff
@@ -186,10 +186,10 @@ public T map(ResultSet rs, StatementContext ctx) throws SQLException {
                             .map(m -> m.getParameterCount() > 0 ? m.getParameters()[0] : null)
                             .orElse(null);
 
-                        QualifiedType type = QualifiedType.of(propertyType(descriptor))
+                        QualifiedType<?> type = QualifiedType.of(propertyType(descriptor))
                             .with(getQualifiers(getter, setter, setterParam));
                         ColumnMapper<?> mapper = ctx.findColumnMapperFor(type)
-                            .orElse((r, n, c) -> r.getObject(n));
+                            .orElse((ColumnMapper) (r, n, c) -> r.getObject(n));
```
Comment (qualidafial) [inline core/src/main/java/org/jdbi/v3/core/mapper/reflect/BeanMapper.java]: This probably wants a `@SuppressWarnings("raw")`

File: core/src/main/java/org/jdbi/v3/core/mapper/reflect/FieldMapper.java [hunk 3]
```diff
@@ -146,10 +146,10 @@ public T map(ResultSet rs, StatementContext ctx) throws SQLException {
 
                     findColumnIndex(paramName, columnNames, columnNameMatchers, () -> debugName(field))
                         .ifPresent(index -> {
-                            QualifiedType type = QualifiedType.of(field.getGenericType())
+                            QualifiedType<?> type = QualifiedType.of(field.getGenericType())
                                 .with(getQualifiers(field));
                             ColumnMapper<?> mapper = ctx.findColumnMapperFor(type)
-                                .orElse((r, n, c) -> r.getObject(n));
+                                .orElse((ColumnMapper) (r, n, c) -> r.getObject(n));
```
Comment (qualidafial) [inline core/src/main/java/org/jdbi/v3/core/mapper/reflect/FieldMapper.java]: Probably wants `@SuppressWarnings("raw")` here too

File: sqlobject/src/main/java/org/jdbi/v3/sqlobject/statement/internal/ResultReturner.java [hunk 4]
```diff
@@ -223,15 +223,15 @@ protected QualifiedType elementType(StatementContext ctx) {
...
         @Override
-        protected QualifiedType elementType(StatementContext ctx) {
+        protected QualifiedType<?> elementType(StatementContext ctx) {
             return elementType;
         }
     }
 
-    static class SingleValueReturner extends ResultReturner {
-        private final QualifiedType returnType;
+    static class SingleValueReturner<T> extends ResultReturner {
```
Comment (qualidafial) [inline sqlobject/src/main/java/org/jdbi/v3/sqlobject/statement/internal/ResultReturner.java]: Any reason for introducing the `<T>` parameter on this and the next returner, but not the others?

Comment (TheRealMarnes) [reply sqlobject/src/main/java/org/jdbi/v3/sqlobject/statement/internal/ResultReturner.java]: the others do `mapType` so become wildcards. Unless I'm mistaken

Comment (TheRealMarnes) [reply sqlobject/src/main/java/org/jdbi/v3/sqlobject/statement/internal/ResultReturner.java]: `ConsumerResultReturner` doesn't get its type from a param, no room for a T
`VoidReturner` idem
`IteratorReturner` does `mapType`
`ResultIteratorReturner` idem
`StreamReturner` idem
`ResultIterableReturner` idem

**Comment by qualidafial** [comment]
@stevenschlansker We're merging to free up other PRs before the release. Let us know if you have any comments and we'll address them separately.

File: core/src/main/java/org/jdbi/v3/core/mapper/ColumnMappers.java [hunk 5]
```diff
@@ -172,19 +172,21 @@ private ColumnMappers register(QualifiedColumnMapperFactory factory) {
      * @return a ColumnMapper for the given type, or empty if no column mapper is registered for the given type.
      */
     @Beta
-    public Optional<ColumnMapper<?>> findFor(QualifiedType type) {
+    public <T> Optional<ColumnMapper<T>> findFor(QualifiedType<T> type) {
         // ConcurrentHashMap can enter an infinite loop on nested computeIfAbsent calls.
         // Since column mappers can decorate other column mappers, we have to populate the cache the old fashioned way.
         // See https://bugs.openjdk.java.net/browse/JDK-8062841, https://bugs.openjdk.java.net/browse/JDK-8142175
-        ColumnMapper<?> cached = cache.get(type);
+        @SuppressWarnings("unchecked")
```
Comment (stevenschlansker) [inline core/src/main/java/org/jdbi/v3/core/mapper/ColumnMappers.java]: Should this `@SuppressWarnings` be on the whole method?  There's a second cast slightly further down.

File: core/src/main/java/org/jdbi/v3/core/mapper/reflect/BeanMapper.java (hunk 2 above)
Comment (stevenschlansker) [reply core/src/main/java/org/jdbi/v3/core/mapper/reflect/BeanMapper.java]: I don't see this as actually resolved.  There's an `unchecked` suppression but not one for `rawtypes`.

File: core/src/main/java/org/jdbi/v3/core/mapper/reflect/FieldMapper.java (hunk 3 above)
Comment (stevenschlansker) [reply core/src/main/java/org/jdbi/v3/core/mapper/reflect/FieldMapper.java]: Ditto, not seeing this as resolved.

File: core/src/test/java/org/jdbi/v3/core/qualifier/TestNVarchar.java [hunk 6]
```diff
@@ -76,18 +76,18 @@ public void sqlStatementBindNVarchar() {
...
             assertThat(
                 handle.select("SELECT name FROM nvarchars ORDER BY id")
-                    .mapTo(new GenericType<String>() {}, NVarchar.class)
+                    .mapTo(QualifiedType.of(new GenericType<String>() {}).with(NVarchar.class))
                     .list())
                 .containsExactly("foo", "bar", "baz", "qux");
 
             List rawList = handle.select("SELECT name FROM nvarchars ORDER BY id")
-                .mapTo(NVARCHAR_STRING)
+                .mapTo((QualifiedType<?>) NVARCHAR_STRING)
```
Comment (stevenschlansker) [inline core/src/test/java/org/jdbi/v3/core/qualifier/TestNVarchar.java]: Why is this cast necessary?  That seems like a real wart that you need to cast it to remove the generic type?

File: core/src/main/java/org/jdbi/v3/core/mapper/ColumnMappers.java (hunk 5 above)
Comment (TheRealMarnes) [reply core/src/main/java/org/jdbi/v3/core/mapper/ColumnMappers.java]: I'm very much against giving suppressions a wide scope, they may end up hiding actual problems or things you _could_ improve if only you noticed them.

File: core/src/test/java/org/jdbi/v3/core/qualifier/TestNVarchar.java (hunk 6 above)
Comment (TheRealMarnes) [reply core/src/test/java/org/jdbi/v3/core/qualifier/TestNVarchar.java]: It's not necessary at all, it's purely to help demonstrate the point of that test, to neglect the generic type.

**Comment by TheRealMarnes** [comment]
Sorry about the merge conflicts. I am trying to wrestle through immutables, I wanna see it merged too. This one can't have caused much though, maybe 2 lines, and I don't think not doing doable work because something else is still open at the same time is a great way to use a ticket-oriented system. If you're blaming me for the quick merge, I didn't decide that or ask for it. And like you said, the intent of the "freeze" wasn't to utterly forbid all other work, especially not work very directly relevant to the release.

**Pull request merged**